### PR TITLE
OGL: Use glBufferData on Mali.

### PR DESCRIPTION
**Pull request by degasus**
tl;dr: This PR speedups dolphin on mobiles with the Mali GPU and ES 3.2
drivers by a factor of 10 by using the method with the biggest overhead.
Please keep care not to buy this shit!

The ARM driver team seems to care very well about their customers. But
bad luck, users and open source developers are *not* their customers. So
even device-independent feature requests are just ignored for *years*:

https://community.arm.com/graphics/f/discussions/4645/gl_ext_buffer_storage-support

The bad point, they neither implement any of the other common ways to
stream dynamic content in unextented GL:
- They just ignore the GL_MAP_UNSYNCHRONIZED_BIT flag
- They don't support on-device buffer updates and just stall with
glBufferSubData

It seems like no benchmark is using any dynamic content - and like no
customer cares about anything but benchmarks, or users...

We have a flag to disable the glBufferSubData way, this PR adds the flag
to also disable the unsychronized mapping way. The second one is
available since their ES 3.2 update, but slow as hell.

So how to continue? The last remaining technical way to stream dynamic
content at all is to alloc a new buffer per draw call with glBufferData.
This is very gross, but still a factor 10 speedup compared to stalling
the GPU. Small tests shows that you can expect another 3-5 times speedup
with EXT_buffer_data, so Mali would be on pair with Adreno here. So if
you have bought such a device unfortunately, please try to make noise on
your vendor forums/support and ask for this extension. If you are going
to buy a new mobile, I'd recormend to avoid *any* mobile with a Mali GPU
in it.

### CODE REVIEW

**Comment by sirdanielk** [comment]
T tested on p9lite (mali t830) with MK DD and i noticed game went faster and smoother in menus, but ingame was almost same , maybe two frames more, finally around 23 FPS. My phone was limited somewhere else. no crashes. I tested with ubershaders disabled.

**Comment by TheZoroark007** [comment]
I have the feeling that this change made Dolphin a bit slower on my Galaxy S8 Mali on Android 8. Before this update , I had between 40-50 FPS in Meele while in a battle, now they won't get above 30 FPS with some drops to 25 FPS.  I will test some other games later.

**Comment by Helios747** [comment]
Blame Mali

**Comment by TheZoroark007** [comment]
Could you somehow make this change optional to select for Mali Devices ? Because it made emulation even slower on the Mali S8 on Android 8 (10FPS less with this build).

**Comment by gahu777** [comment]
Also have problem with this update on my mate9 on gl version 3.2 :/

**Comment by gahu777** [comment]
This update slow emulation speed in every game

**Comment by loganmc10** [comment]
OpenGL ES still allows client side arrays, there is no need to use glBufferSubData/glBufferData/glMapBufferRange at all on these devices.

**Comment by degasus** [comment]
@loganmc10 indeed, there is no need, but the required code changes for client side arrays are bad. It would prohebit to use VAO as well.
